### PR TITLE
Relation support to N=15

### PR DESCRIPTION
**Pull request by stephentetley**
This PR is work supporting tuples and relations to arity N=15.

Tuple instances for the classes `ToString`, `Eq`, `Order` and `Boxable` have been implemented up to 15.

The `Guard` and `App` datatypes in the new solver in the new solver have been extended up to 15. The families of `projectIntoX` and `factsX` functions have been implemented up to 15.

### CODE REVIEW

**Comment by magnus-madsen** [comment]
Thanks Stephen. I will take a look soon. 

Can I ask you for favor? Can you run `--Xbenchmark-throughput --json` on master and on this PR and post the results.
(It takes a few minutes)

I don't expect this should have too big of an impact on performance, but I just want to be sure before merging.

**Comment by stephentetley** [comment]
# master

"C:\Program Files\AdoptOpenJDK\jre-15.0.2.7-hotspot\bin\java.exe" "-javaagent:C:\Program Files\JetBrains\IntelliJ IDEA Community Edition 2020.2\lib\idea_rt.jar=57137:C:\Program Files\JetBrains\IntelliJ IDEA Community Edition 2020.2\bin" -Dfile.encoding=UTF-8 -classpath E:\coding\flix\temp\master\out\production\master;C:\Users\stephen\.ivy2\cache\org.scala-lang\scala-library\jars\scala-library-2.13.5.jar;C:\Users\stephen\.ivy2\cache\org.scala-lang\scala-reflect\jars\scala-reflect-2.13.5.jar;C:\Users\stephen\.ivy2\cache\org.scala-lang\scala-library\srcs\scala-library-2.13.5-sources.jar;E:\coding\flix\temp\master\lib\com.github.scopt-3.7.1.jar;E:\coding\flix\temp\master\lib\com.microsoft.z3.jar;E:\coding\flix\temp\master\lib\flix-runtime.jar;E:\coding\flix\temp\master\lib\org.java_websocket-1.3.9.jar;E:\coding\flix\temp\master\lib\org.jline-3.5.1.jar;E:\coding\flix\temp\master\lib\org.json4s-ast-3.5.5.jar;E:\coding\flix\temp\master\lib\org.json4s-core-3.5.5.jar;E:\coding\flix\temp\master\lib\org.json4s-native-3.5.5.jar;E:\coding\flix\temp\master\lib\org.objectweb.asm-all-5.2.jar;E:\coding\flix\temp\master\lib\org.parboiled-2.2.1.jar;E:\coding\flix\temp\master\lib\org.scalactic-3.0.8.jar;E:\coding\flix\temp\master\lib\org.scalatest-3.0.8.jar;E:\coding\flix\temp\master\lib\scala.collection.parallel-0.2.0.jar;E:\coding\flix\temp\master\lib\scala.xml-1.2.0.jar;E:\coding\flix\temp\master\lib\shapeless-2.3.3.jar ca.uwaterloo.flix.Main --Xbenchmark-throughput --json
{
  "threads":4,
  "lines":42678,
  "iterations":15,
  "throughput":{
    "min":1799,
    "max":7882,
    "avg":6405,
    "median":6992
  }
}

Process finished with exit code 0

# n=15

"C:\Program Files\AdoptOpenJDK\jre-15.0.2.7-hotspot\bin\java.exe" "-javaagent:C:\Program Files\JetBrains\IntelliJ IDEA Community Edition 2020.2\lib\idea_rt.jar=57312:C:\Program Files\JetBrains\IntelliJ IDEA Community Edition 2020.2\bin" -Dfile.encoding=UTF-8 -classpath E:\coding\flix\temp\n15\out\production\n15;C:\Users\stephen\.ivy2\cache\org.scala-lang\scala-library\jars\scala-library-2.13.5.jar;C:\Users\stephen\.ivy2\cache\org.scala-lang\scala-reflect\jars\scala-reflect-2.13.5.jar;C:\Users\stephen\.ivy2\cache\org.scala-lang\scala-library\srcs\scala-library-2.13.5-sources.jar;E:\coding\flix\temp\n15\lib\com.github.scopt-3.7.1.jar;E:\coding\flix\temp\n15\lib\com.microsoft.z3.jar;E:\coding\flix\temp\n15\lib\flix-runtime.jar;E:\coding\flix\temp\n15\lib\org.java_websocket-1.3.9.jar;E:\coding\flix\temp\n15\lib\org.jline-3.5.1.jar;E:\coding\flix\temp\n15\lib\org.json4s-ast-3.5.5.jar;E:\coding\flix\temp\n15\lib\org.json4s-core-3.5.5.jar;E:\coding\flix\temp\n15\lib\org.json4s-native-3.5.5.jar;E:\coding\flix\temp\n15\lib\org.objectweb.asm-all-5.2.jar;E:\coding\flix\temp\n15\lib\org.parboiled-2.2.1.jar;E:\coding\flix\temp\n15\lib\org.scalactic-3.0.8.jar;E:\coding\flix\temp\n15\lib\org.scalatest-3.0.8.jar;E:\coding\flix\temp\n15\lib\scala.collection.parallel-0.2.0.jar;E:\coding\flix\temp\n15\lib\scala.xml-1.2.0.jar;E:\coding\flix\temp\n15\lib\shapeless-2.3.3.jar ca.uwaterloo.flix.Main --Xbenchmark-throughput --json
{
  "threads":4,
  "lines":44412,
  "iterations":15,
  "throughput":{
    "min":1744,
    "max":4758,
    "avg":4166,
    "median":4368
  }
}

Process finished with exit code 0

**Comment by magnus-madsen** [comment]
It does seem there is a worrying slowdown (avg 6405 vs 4166). My plan is to:

- Replicate on my machine.
- See if this is related to a bottleneck in `Monomorph` (as I expect).
- If it is then we can still merge it now and I will open a ticket so Matt can look into it.

Essentially I am OK with a temporary slowdown of master as long as we have a line of attack. 

Stephen, does this PR make your program run? Or are you blocked by exceptions from `Lowering.scala`? (in which case I could fix that in master and then you could use this PR until we figure out what is going on). I am just trying to ensure that you are not too blocked.

Ill look into these issues later today or tomorrow.

**Comment by stephentetley** [comment]
Thanks Magnus, my program runs without changes to `Lowering.scala`.

**Comment by magnus-madsen** [comment]
So these are the results:

Master:
```
{
  "threads":6,
  "lines":42678,
  "iterations":15,
  "throughput":{
    "min":4789,
    "max":21511,
    "avg":17848,
    "median":20027
  }
}
====================== Flix Compiler Phases ======================

Best runtime per phase (low to high):
  Documentor                       0,0 ms (00,0%)
  Reader                           0,4 ms (00,0%)
  Instances                        1,3 ms (00,1%)
  Safety                           4,4 ms (00,2%)
  Tailrec                          4,5 ms (00,2%)
  VarNumbering                     7,9 ms (00,4%)
  Lowering                         8,4 ms (00,4%)
  Stratifier                       9,6 ms (00,5%)
  Optimizer                       11,0 ms (00,6%)
  LambdaLift                      11,8 ms (00,6%)
  Simplifier                      12,5 ms (00,6%)
  Weeder                          13,4 ms (00,7%)
  Eraser                          19,3 ms (01,0%)
  PatternExhaustiveness           20,3 ms (01,0%)
  Redundancy                      26,3 ms (01,3%)
  TreeShaker                      28,0 ms (01,4%)
  ClosureConv                     29,9 ms (01,5%)
  Finalize                        31,6 ms (01,6%)
  Namer                           40,4 ms (02,0%)
  Resolver                        86,5 ms (04,4%)
  Parser                          98,7 ms (05,0%)
  Monomorph                      154,7 ms (07,8%)
  Typer                          568,1 ms (28,8%)
  JvmBackend                     785,1 ms (39,8%)

Finished 15 iterations on 42.678 lines of code in 40 seconds.
```

PR:
```
{
  "threads":6,
  "lines":44412,
  "iterations":15,
  "throughput":{
    "min":5317,
    "max":13627,
    "avg":11783,
    "median":12284
  }
}

====================== Flix Compiler Phases ======================

Best runtime per phase (low to high):
  Documentor                       0,0 ms (00,0%)
  Reader                           0,8 ms (00,0%)
  Instances                        1,7 ms (00,1%)
  Safety                           4,3 ms (00,1%)
  Tailrec                          4,4 ms (00,1%)
  VarNumbering                     7,8 ms (00,2%)
  Simplifier                       9,2 ms (00,3%)
  Stratifier                       9,5 ms (00,3%)
  Lowering                        10,0 ms (00,3%)
  Optimizer                       10,9 ms (00,3%)
  Weeder                          11,0 ms (00,3%)
  LambdaLift                      11,7 ms (00,4%)
  Eraser                          19,1 ms (00,6%)
  PatternExhaustiveness           22,0 ms (00,7%)
  TreeShaker                      26,4 ms (00,8%)
  Redundancy                      29,0 ms (00,9%)
  ClosureConv                     29,3 ms (00,9%)
  Finalize                        34,0 ms (01,1%)
  Namer                           40,4 ms (01,3%)
  Resolver                        90,2 ms (02,8%)
  Parser                          95,8 ms (03,0%)
  Monomorph                      169,2 ms (05,3%)
  JvmBackend                     758,9 ms (23,7%)
  Typer                          1809,5 ms (56,5%)

Finished 15 iterations on 44.412 lines of code in 58 seconds.
```

The performance issue is real and is here:

```
  Monomorph                      169,2 ms (05,3%)
  JvmBackend                     758,9 ms (23,7%)
  Typer                          1809,5 ms (56,5%)
```

compare:


```
  Monomorph                      154,7 ms (07,8%)
  Typer                          568,1 ms (28,8%)
  JvmBackend                     785,1 ms (39,8%)
```

Unfortunately the issue is not in `Monomorph` as expected (although there is also a slow down there), but rather in `Typer`.

**Comment by magnus-madsen** [comment]
Note that the slow down is quiet dramatic `568,1 ms -> 1809,5 ms`. Its a bit shocking.

What I'd like to do is two things:

First, kindly ask @mlutze if he has any idea why this would happen for these high arity functions? (While there is a performance issue in Monomorph ~1.2x slowdown, Typer has 3.0x slow down. Any idea why?

Second, the PR has N=15 for "everything", but there are really two numbers at stake:

- N for tuples (e.g. Boxable, Eq, Order, ToString) determines the *maximum arity of tuples that can project into/out of Datalog values*.
- M for `GuardX` and `AppX` determines the *maximum number of variables in a rule* (ignoring wildcards). 

Stephen, for  your purposes you may only need N = 15, but M=10 might still work. I am going to see if that has any performance impact.

**Comment by magnus-madsen** [comment]
I spent some time on this. 

***TLDR:***

Stephen if you could update the PR (or create a new) to *only* add Eq, Order, ToString, and Boxable instances for tuples up to N=15 then I think that can be merged without performance concerns. Moreover, for your current use cases the limitations of at most 5 distinct variables per expression in a if-guard or head expression term is not a problem. (I can expand on this, but since you are not seeing exceptions from Lowering this is the case.)

Going forward, I create all of these tickets: https://github.com/flix/flix/issues/1888 https://github.com/flix/flix/issues/1886 https://github.com/flix/flix/issues/1887

**Comment by magnus-madsen** [comment]
With only the proposed changes this is still allowed:

```
A(a,b,c,d,e,f,g,h,j,k,l,m,o,p,q,r,s,t,u,v,w,x,y,z) :- B(a,b,c,d,e,f,g,h,j,k,l,m,o,p,q,r,s,t,u,v,w,x,y,z), C(a,b,c,d,e,f,g,h,j,k,l,m,o,p,q,r,s,t,u,v,w,x,y,z)
```

The limitation is that you can "only" have:

```
Q(a+b+c+d+e, u+v+w+x+y+z) :- B(a,b,c,d,e,f,g,h,j,k,l,m,o,p,q,r,s,t,u,v,w,x,y,z), C(a,b,c,d,e,f,g,h,j,k,l,m,o,p,q,r,s,t,u,v,w,x,y,z)
```

(as an example.)

**Comment by stephentetley** [comment]
Thanks Magnus, I'll update the PR this evening after work.

**Comment by stephentetley** [comment]
I've created an alternative PR with just the instances...

Up to n=15 tuple instances of Boxable, Eq, Order and ToString #1891

**Comment by magnus-madsen** [comment]
I am closing this. We now support relations up to N=15 and projecting into/out of these.

What we don't support is *single terms with more than 5 variables in each term*.
